src/internal/database/tenant.ts

```diff
@@ -250,11 +262,169 @@ export async function listenForTenantUpdate(pubSub: PubSubAdapter): Promise<void
     tenantConfigCache.delete(cacheKey)
   })
 
+  await pubSub.subscribe(TENANTS_JWKS_UPDATE_CHANNEL, (cacheKey) => {
+    tenantJwksConfigCache.delete(cacheKey)
+  })
+
   await pubSub.subscribe(TENANTS_S3_CREDENTIALS_UPDATE_CHANNEL, (cacheKey) => {
     tenantS3CredentialsCache.delete(cacheKey)
   })
 }
 
+/**
+ * Gets a list of all tenants that do not have a signing url associated
+ */
+export async function* listTenantsMissingUrlSigningJwk(
+  signal: AbortSignal,
+  batchSize = 200
+): AsyncGenerator<string[]> {
+  let lastCursor = 0
+
+  while (!signal.aborted) {
+    const data = await multitenantKnex('tenants')
+      .select('id', 'cursor_id')
+      .where('cursor_id', '>', lastCursor)
+      .whereNotExists(function () {
+        this.select(1)
+          .from('tenants_jwks')
+          .whereRaw('tenants_jwks.tenant_id = tenants.id')
+          .andWhere('tenants_jwks.kind', JWK_KIND_STORAGE_URL_SIGNING)
+          .andWhere('tenants_jwks.active', true)
+      })
+      .orderBy('cursor_id', 'asc')
+      .limit(batchSize)
+
+    if (data.length === 0) {
+      break
+    }
+
+    lastCursor = data[data.length - 1].cursor_id
+    yield data.map((tenant) => tenant.id)
+  }
+}
+
+/**
+ * Generates a new URL signing JWK and stores it in the database if one does not already exist.
+ * Only one active url signing jwk can exist, this function is idempotent and will create a new entry or return the kid of the existing
+ * @param tenantId
+ * @param trx optional transaction to add the jwk within
+ */
+export async function generateUrlSigningJwk(
+  tenantId: string,
+  trx?: Knex.Transaction
+): Promise<{ kid: string }> {
+  const query = trx || multitenantKnex
+
+  return await query.transaction(async (transaction) => {
+    const activeJwk = await transaction('tenants_jwks')
```

Comment:
as discussed, we need a unique index instead of find for update to guarantee uniqueness and idempotency